iavl/iavl_proof2.go

```diff
@@ -167,52 +171,266 @@ func (proof *KeyNotExistsProof) Verify(key []byte, root []byte) error {
+		}
+	}
+
+	// Verify that all paths are adjacent to one another.
+	if err := proof.verifyPathAdjacency(); err != nil {
+		return err
+	}
+
+	// If proof.PathToKeys is empty, it means we have an empty range. This range
+	// can be between keys, or outside of the range of existing keys.
+	if len(proof.PathToKeys) == 0 {
+		if proof.LeftPath == nil {
+			// Range is outisde and to the left of existing keys.
+			if !proof.RightPath.isLeftmost() {
+				return errors.New("right path is not leftmost")
+			}
+		} else if proof.RightPath == nil {
+			// Range is outisde and to the right of existing keys.
+			if !proof.LeftPath.isRightmost() {
+				return errors.New("left path is not rightmost")
+			}
+		}
+	} else {
+		// If we've reached this point, it means our range isn't empty, and we have
+		// a list of keys.
+		for i, path := range proof.PathToKeys {
+			leafNode := IAVLProofLeafNode{KeyBytes: keys[i], ValueBytes: values[i]}
+			if err := path.verify(leafNode, root); err != nil {
+				return errors.Wrap(err, "failed to verify inner path")
+			}
+		}
+	}
+	return nil
+}
+
+func (node *IAVLNode) pathToKey(t *IAVLTree, key []byte) (*PathToKey, []byte, error) {
+	path := &PathToKey{}
+	val, err := node._pathToKey(t, key, path)
+	return path, val, err
+}
+func (node *IAVLNode) _pathToKey(t *IAVLTree, key []byte, path *PathToKey) ([]byte, error) {
 	if node.height == 0 {
 		if bytes.Compare(node.key, key) == 0 {
-			proof.LeafHash = node.hash
+			path.LeafHash = node.hash
 			return node.value, nil
 		}
 		return nil, errors.New("key does not exist")
 	}
 
 	if bytes.Compare(key, node.key) < 0 {
-		if value, err := node.getLeftNode(t).constructKeyExistsProof(t, key, proof); err == nil {
+		if value, err := node.getLeftNode(t)._pathToKey(t, key, path); err == nil {
 			branch := IAVLProofInnerNode{
 				Height: node.height,
 				Size:   node.size,
 				Left:   nil,
 				Right:  node.getRightNode(t).hash,
 			}
-			proof.InnerNodes = append(proof.InnerNodes, branch)
+			path.InnerNodes = append(path.InnerNodes, branch)
 			return value, nil
 		}
 		return nil, errors.New("key does not exist")
 	}
 
-	if value, err := node.getRightNode(t).constructKeyExistsProof(t, key, proof); err == nil {
+	if value, err := node.getRightNode(t)._pathToKey(t, key, path); err == nil {
 		branch := IAVLProofInnerNode{
 			Height: node.height,
 			Size:   node.size,
 			Left:   node.getLeftNode(t).hash,
 			Right:  nil,
 		}
-		proof.InnerNodes = append(proof.InnerNodes, branch)
+		path.InnerNodes = append(path.InnerNodes, branch)
 		return value, nil
 	}
 	return nil, errors.New("key does not exist")
 }
 
-func (node *IAVLNode) constructKeyNotExistsProof(t *IAVLTree, key []byte, proof *KeyNotExistsProof) error {
+func (node *IAVLNode) constructKeyRangeProof(
+	t *IAVLTree, keyStart, keyEnd []byte, limit int, rangeProof *KeyRangeProof,
+) (
+	keys [][]byte, values [][]byte, err error,
+) {
+	keys, values = [][]byte{}, [][]byte{}
```

Comment:
Yeah, only difference is it'll be `nil` if empty, not `[]byte{}`, but I've changed the test to match. f203deed703a0e859ce8be0dbaee40b5ac68d2ef